### ISSUE TITLE
Nokia private config mode and prompt pattern - error when using private mode

### ISSUE THREAD

**Issue by Menotomy** [ktbyers/netmiko#1738]
Currently the config_mode() method for Nokia uses "edit-config exclusive" then searches for the pattern "(ex) [" as that is the prompt when entering that mode (this is when using MD-CLI).

I'm trying to use edit-config private so I'm not stepping on any other edits. When using the "edit-config private" mode in MD-CLI, the prompt is "(pr) [". However, even when updating the pattern to use pr, it fails to enter config mode. 

Example prompt:
>[]
> A:user@Nokia# edit-config private
> INFO: CLI #2070: Entering private configuration mode
> INFO: CLI #2061: Uncommitted changes are discarded on configuration mode exit
> 
> (pr)[]
> A:user@Nokia#

The line that calls it looks like this:
`
connection.config_mode(config_command='edit-config private', pattern='\(pr\)\[')
`

_Am I formatting the pattern properly?_ 

The API doc is a little confusing because it shows the format as `'\\(pr\\)\\['`, which didn't work either. Also in the source code it uses a literal string `r'\(pr\)\['`.

Here is the error:

> /usr/local/lib/python3.7/site-packages/netmiko/base_connection.py in config_mode(self=<netmiko.nokia.nokia_sros_ssh.NokiaSrosSSH object>, config_command='edit-config private', pattern=r'\(pr\)\[')
> --
> 1623                 output += self.read_until_pattern(pattern=pattern)
> 1624             if not self.check_config_mode():
> => 1625                 raise ValueError("Failed to enter configuration mode.")
> 1626         return output
> 1627
> builtin ValueError = <class 'ValueError'>

**Comment by ktbyers** [comment]
Adding @hellt

**Comment by hellt** [comment]
thanks @Menotomy for bringing this up!

The Nokia SR OS was designed to work with exclusive mode only. The reason for that was to default to the "safe" mode where the automation system (a script) would _lock_ the device by entering in the exclusive mode and guarantee that noone will alter the configuration state in a concurrent session.

Making the changes to the private candidates do make sense, just not that much in the automation scenarios. You still loose your uncommited changes upon leaving the private configuration mode, but you expose yourself for a situation when someone will change the piece in another configuration session that you did not expect to be changed.

It could be that you wanted to edit the global configuration, that one will persist when you disconnect from the node?

In any case, currently your only way around is to use _raw_ `send_command` method, that doesn't care about the config modes and just sends the data on the wire. Example:

```
nc.send_command("edit-config private")
nc.send_command(
    "/configure router interface edit_private ipv4 primary address 1.1.1.10 prefix-length 32 "
)
```

**Comment by Menotomy** [comment]
Thanks for the reply. 

Interestingly in my company we were looking at using a private mode scenario for all of our vendors in production automation. The thought being there could be a concurrent config push, though that requires coordination by the northbound systems potentially. Also, the entire configuration isn't automated, just independent pieces making overlap less likely. The behavior would be more like using NETCONF to push a config into a candidate store then committing it. 

This particular example is for a lab and the purpose is when the company adds a new subnet to a office area or VPN, a coworker can use this script via CGI (old, I know, but it's simple and it works) to enter a subnet and configure a couple dozen routers. My intention of using configure private was to do that in a non intrusive way during daytime hours when someone may be in global config mode. 

If exclusive is the intention of Netmiko, I'm fine with that. Honestly, it's for a lab so who cares. If it can't go into that mode then someone will have to add it manually, not a big deal. 

I used send_command_set so it would iterate through a Python list that's built by a loop that reads a file with not only the static route but also a firewall prefix-set and replaces variables for the subnet and the mask. If it were just one command then I agree that single line would work, but I felt it was cleaner to use the _set version. 

Probably more info than you wanted...

**Comment by hellt** [comment]
> The behavior would be more like using NETCONF to push a config into a candidate store then committing it.

That is indeed after the NETCONF we modelled the SR OS driver. With NC a system _should_ use the `<lock>` on the candidate first before making the changes to it and `<unlock>` after.
That is what we do with `edit-config exclusive`.

**Comment by ktbyers** [comment]
In Netmiko, you could probably use a pattern similar to this:

```python
nc.send_command("edit-config private", expect_string=r"(pr)")
nc.send_config_set(config_commands, enter_config_mode=False, exit_config_mode=False)
# then need to manually exit config mode...
```

I somewhat suspect that would work, but you would need to test it.

**Comment by Menotomy** [comment]
@hellt 
Yes, exclusive would be similar to using lock. Internally we were discussing if we want to do a lock when we have a system push a lot of configs, like several different interfaces to ready them for testing, and we know they won't overlap. I do understand how the driver was constructed though, not trying to argue that. 

@ktbyers 
I tried that but I'm wondering if something else would be needed at commit. Looking at the output it doesn't commit, it just exits. I looked at the commit() code in your API doc and it looks like it still checks for (ex), without that I think it just exits. I didn't see a way to pass a different pattern. 

However, it did successfully enter config mode, so your suggestion worked except for the above commit part. 

Since this is closed I don't expect any development, just noting it for future reference if someone else sees it.

> INFO: CLI #2070: Entering private configuration mode 
> INFO: CLI #2061: Uncommitted changes are discarded on configuration mode exit 
> INFO: CLI #2075: Other configuration sessions are active 
> /configure filter { match-list ip-prefix-list "LIST-V4" prefix 1.1.1.0/22 }
>  *(pr)[configure] 
> A:user@RTR2# /configure router "management" { static-routes route 1.1.1.0/22 route-type unicast next-hop "1.1.1.1" } 
> *(pr)[configure]
>  A:user@RTR2# exit all 
> *(pr)[] 
> A:user@RTR2#

Here is the code as you suggested:
```
print(connection.send_command('edit-config private', expect_string=r'(pr)'))
print(connection.send_config_set(config_send, enter_config_mode=False, exit_config_mode=False))
print(connection.commit())
```

The code API:
https://ktbyers.github.io/netmiko/docs/netmiko/nokia/index.html#netmiko.nokia.NokiaSrosSSH.commit
